components/event-publisher/event-output-adapters/org.wso2.carbon.event.output.adapter.email/src/main/java/org/wso2/carbon/event/output/adapter/email/EmailEventAdapter.java

```diff
@@ -375,10 +430,56 @@ public void run() {
                 EmailEventAdapterUtil.logAndDrop(eventAdapterConfiguration.getName(), message,
                         "Error in message format", e, log, tenantId);
             } catch (Exception e) {
-                EventAdapterUtil.logAndDrop(eventAdapterConfiguration.getName(), message, "Error sending email to '" + to + "'", e, log, tenantId);
+                EventAdapterUtil.logAndDrop(eventAdapterConfiguration.getName(), message,
+                        "Error sending email to '" + to + "'", e, log, tenantId);
+            }
+        }
+    }
+
+    private void sendWithRetry(Message message) throws MessagingException {
+
+
+        try {
+            Transport.send(message);
+        } catch (AuthenticationFailedException exception) {
+            // Retry only for CLIENT_CREDENTIAL authentication type.
+            if (!CLIENT_CREDENTIAL.equalsIgnoreCase(globalProperties.get(ADAPTER_EMAIL_AUTH_TYPE))) {
+                throw new AuthenticationFailedException();
             }
+            handleRetry(message);
+            log.warn("Authentication failed, attempting token refresh and retry...", exception);
         }
+    }
 
+    private void handleRetry(Message message) throws MessagingException {
+
+        int attempts = 0;
+
+        while (attempts < MAX_RETRY_ATTEMPTS) {
+            SMTPTransport transport;
+            attempts++;
+
+            try {
+                transport = (SMTPTransport) session.getTransport(SMTP_PROTOCOL);
+                smtpPassword = getAccessToken(clientId, clientSecret,
+                        globalProperties.get(ADAPTER_EMAIL_TOKEN_ENDPOINT),
+                        globalProperties.get(ADAPTER_EMAIL_SCOPES));
+                transport.connect(globalProperties.get(MAIL_SMTP_HOST), smtpUsername, smtpPassword);
+
+                transport.sendMessage(message, message.getAllRecipients());
+                if (log.isDebugEnabled()) {
+                    log.debug("Mail sent to the EmailID " + Arrays.toString(message.getAllRecipients()) +
+                            " Successfully with retry.");
+                }
+                return;
+
+            } catch (Exception e) {
+                log.warn("Authentication failed, attempting token refresh and retry (attempt " + attempts + ")", e);
```

Comment:
Let's not print e, here. For the debugging purpose may be can add 'e' with a debug log.